python/packages/isce3/math/offsets_polyfit.py

```diff
@@ -0,0 +1,508 @@
+
+    Returns
+    -------
+    int
+        The number of unique coefficients in the polynomial.
+    """
+    return (degree + 1) * (degree + 2) // 2
+
+
+def normalize(x, xmin, xmax):
+    """
+    Linearly normalize values to the range [0, 1].
+
+    Parameters
+    ----------
+    x : array_like
+        Input array or scalar to be normalized.
+    xmin : float or int
+        Minimum value of the input range.
+    xmax : float or int
+        Maximum value of the input range.
+
+    Returns
+    -------
+    ndarray
+        Normalized array with values scaled to the interval [0, 1].
+        If `xmax` equals `xmin`, returns an array of zeros with the same shape as `x`.
+    """
+    if xmax == xmin:
+        return np.zeros_like(x, dtype=float)
+    return (x - xmin) / (xmax - xmin)
+
+
+def build_design_matrix(lines_n, pixels_n, degree: int):
+    """
+    Build the design matrix for a 2D polynomial regression.
+
+    Each row of the design matrix corresponds to one observation, with
+    columns containing all polynomial terms up to the specified total degree.
+    Specifically, for each observation `i`:
+
+        A[i, :] = [l^(j-k) * p^k]  for all  0 <= j <= degree,  0 <= k <= j
+
+    where `l` and `p` are the normalized line and pixel coordinates.
+
+    Parameters
+    ----------
+    lines_n : array_like
+        Normalized line coordinates (independent variable 1).
+    pixels_n : array_like
+        Normalized pixel coordinates (independent variable 2).
+    degree : int
+        The maximum total degree of the 2D polynomial.
+
+    Returns
+    -------
+    A : ndarray of shape (N, M)
+        The design matrix, where `N` is the number of observations and
+        `M` is the number of polynomial coefficients, given by:
+
+        M = (degree + 1) * (degree + 2) / 2
+    """
+    N = lines_n.size
+    A = np.zeros((N, ncoeffs(degree)), dtype=float)
+    idx = 0
+    for j in range(degree + 1):
+        for k in range(j + 1):
+            A[:, idx] = (lines_n ** (j - k)) * (pixels_n ** k)
+            idx += 1
+    return A
+
+
+def cholesky_solve(N_mat, rhs):
+    """
+    Solve a symmetric positive-definite linear system using Cholesky decomposition.
+
+    This function solves the system:
+
+        N_mat * x = rhs
+
+    where `N_mat` is a symmetric positive-definite matrix, by performing
+    a Cholesky decomposition `N_mat = L @ L.T` and solving the two
+    triangular systems `L y = rhs` and `L.T x = y`.
+
+    Parameters
+    ----------
+    N_mat : (M, M) array_like
+        Symmetric positive-definite coefficient matrix.
+    rhs : (M,) or (M, K) array_like
+        Right-hand side vector or matrix.
+
+    Returns
+    -------
+    x : ndarray
+        Solution vector or matrix satisfying `N_mat @ x = rhs`.
+    L : ndarray
+        Lower-triangular Cholesky factor of `N_mat`.
+    """
+
+    L = np.linalg.cholesky(N_mat)
+    y = np.linalg.solve(L, rhs)
+    x = np.linalg.solve(L.T, y)
```

Comment:
The numpy solver doesn't claim to check for any structure in the matrix, so IMO we should use scipy routines since we already depend on scipy.
```suggestion
    L, is_lower = scipy.linalg.cholesky(N_mat, lower=True)
    x = scipy.linalg.cho_solve((L, is_lower), rhs)
```

Reply:
very good suggestion